docs/sources/mimir/operators-guide/mimir-runbooks/_index.md

```diff
@@ -592,18 +592,37 @@ How to **investigate**:
-        - For each failed compaction job
-          - Pick one result block (doesn't matter which)
-          - Find source blocks for the compaction job: search for `msg="compact blocks"` and a mention of the result block ID.
-          - Mark the source blocks for no compaction (in this example the object storage backend is GCS):
-            ```
-            ./tools/markblocks/markblocks -backend gcs -gcs.bucket-name <bucket> -mark no-compact -tenant <tenant-id> -details "Leading to out-of-order chunks when compacting with other blocks" <block-1> <block-2>...
-            ```
+  - Out-of-order chunks
+    - **How to detect**: Search compactor logs for `invalid result block` and `out-of-order chunks`.
+    - This is caused by a bug in the ingester - see [mimir#1537](https://github.com/grafana/mimir/issues/1537). Ingesters upload blocks where the MinT and MaxT of some chunks don't match the first and last samples in the chunk. When the faulty chunks' MinT and MaxT overlap with other chunks, the compactor merges the chunks. Because one chunk's MinT and MaxT are incorrect the merge may be performed incorrectly, leading to OoO samples.
+    - **How to mitigate**: Mark the faulty blocks to avoid compacting them in the future:
+      - Find all affected compaction groups in the compactor logs. You will find them as `invalid result block /data/compact/<compaction_group>/<result_block>`.
+      - For each failed compaction job
+        - Pick one result block (doesn't matter which)
+        - Find source blocks for the compaction job: search for `msg="compact blocks"` and a mention of the result block ID.
+        - Mark the source blocks for no compaction (in this example the object storage backend is GCS):
+          ```
+          ./tools/markblocks/markblocks -backend gcs -gcs.bucket-name <bucket> -mark no-compact -tenant <tenant-id> -details "Leading to out-of-order chunks when compacting with other blocks" <block-1> <block-2>...
+          ```
+
+Another possible cause of this alert is if compaction has never been running successfully, potentially for a long time. This can be because the alert either was not seen, misconfigured or not installed at all for some time.
+
+This can also result in long-term queries failing as the store-gateways fail to handle the much larger number of smaller blocks than expected.
+
+How to **investigate**:
+
+- Check the [Compactor Dashboard]({{< relref "../monitor-grafana-mimir/dashboards/compactor/index.md" >}})
+  - Compactor has fallen behind:
+    - **How to detect**:
+      - Check the `Last successful run per-compactor replica` panel - are there recent runs in the last 6-12 hours?
+      - Also check the `Average blocks / tenant` panel - what is the count? A tenant would have one 24h block per day plus smaller blocks to be compacted if all is well. Given that, <1200 blocks could be normal. Thousands and thousands of blocks is not normal.
```

Comment:
In this panel the long-term trend is more important than the current number. Maybe suggest to view the dashboard for 7 days and see if the number is steadily increasing.

Reply:
I think both are valid - you definitely would like to know if it's steadily increasing but in a scenario where you've only just looked at the panel for the first time, if it shows you're at ~50k blocks, well even without a trend that shows you something is wrong, no?

Reply:
I can change the language to suit both easy enough, though.

Reply:
> A tenant would have one 24h block per day plus smaller blocks to be compacted if all is well.

that's not true for the split&merge compactor. Maybe reword to include that detail too

> Thousands and thousands of blocks is not normal.

Maybe cap this at 50K blocks? or 100K? We have a tenant with 7K and I can image the same tenant cluster with a few years' retention will have a multiple of 7K.

Reply:
> that's not true for the split&merge compactor. Maybe reword to include that detail too

How would you ballpark it for split&merge? Afaik once merging has occurred you do still end up with 1 24h block per day, just you may have a larger quantity of smaller blocks that are still undergoing split-and-merge, if that's incorrect let me know :)

> Maybe cap this at 50K blocks? or 100K? We have a tenant with 7K and I can image the same tenant cluster with a few years' retention will have a multiple of 7K.

I debated what number to use here, which is why I tried to add the age of tenant above, if 50k is a safer "silly" number, happy to go with that.

Reply:
> How would you ballpark it for split&merge? Afaik once merging has occurred you do still end up with 1 24h block per day, just you may have a larger quantity of smaller blocks that are still undergoing split-and-merge, if that's incorrect let me know :)

you get 1 24hr block for each `-compactor.split-and-merge-shards`. The idea is that each block contains series whose labels hash modulo `-compactor.split-and-merge-shards` is the same.